### PR TITLE
48019022: Semi Supervised Multi-layer Graph Neural Network for Node Classification of the Facebook Large Page-Page Data

### PR DESCRIPTION
**Pull request by anthonylastnamengo**
Normal Difficulty. This project tests multiple deep learning graph neural networks, including GCNs, GATs, GraphSAGE, and SGC, to classify nodes of the Facebook dataset into 4 classes. The final model (GAT) achieved a test accuracy of over 98%.

The following files are included in the repository:
- train.py: Training and evaluation logic wrapped in a main training loop.
- modules.py: Contains various GNN architectures, defines layers and forward passes.
- dataset.py: Custom dataloader for the Facebook graph data.
- predict.py: Script for running inference on the dataset. 
- plotting.py: Script for plotting the TSNE embeddings for a select model.
- wandb_config.py: Script for initialising Weights and Biases logging.
- utils.py: (Unused)

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Normal

Readme Overall: Good

- Project Overview: Good
- Model Documented/Explained: Good
- Data + Pre-processing: Not done. Used 128 embedding as model input. This is ok.
- Training / Loss Curve(s): Good.
- Result Demonstration: Good
- References: Good
- Setup/Reproducibility/Requirements: Good

Functionality:

- Multi-layer GCN: Good, 2 layers. Uses external library (torch_geometric.nn GCNConv).  Also implemented GAT/SAGE/SGC  GNN variants. Good!
- Semi supervised: Good
- Accuracy: Good.
- TSNE or UMAP: Good. Class discrimination good.
- Discussion: Good

Code: 

- Consistent with Results: Yes.
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good
- Reusable Modular code/Design: Good

Other Comments:
Has patience based best model selection - Good.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Hard coded hyper parameters could be avoided in GATModelBasic (heads can be a parameter)</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-6</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (gitignore). Please update for merge, doesn't affect grade.
